### ISSUE TITLE
ObjectDisposedException should be thrown on access to public members of IDisposables

### ISSUE THREAD

**Issue by dodexahedron** [gui-cs/Terminal.Gui#3105]
What it says on the tin.

While this isn't necessarily a _bug_, per se, it is a design recommendation that helps consumers significantly, if they have use-after-free bugs in their code, accessing an object that has been disposed.

From [Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern):

> DO throw an [ObjectDisposedException](https://learn.microsoft.com/en-us/dotnet/api/system.objectdisposedexception) from any member that cannot be used after the object has been disposed of.

In some unit tests in TGD, I ran into almost all as NullReferenceExceptions in this situation, but also a few others that could possibly indicate potentially undesirable side-effects.

Throwing ObjectDisposedException is a big help to consumers, as the exception itself and the stack trace are muuuuuch more helpful than the downstream issues caused by the access to the invalid object and happen right away, which is also a big help in debugging.

**Comment by dodexahedron** [comment]
.net 8 has a shortcut method for this check, to avoid a bunch of boilerplate if blocks:

`ObjectDisposedException.ThrowIf(bool condition, object instance)`

https://learn.microsoft.com/en-us/dotnet/api/system.objectdisposedexception.throwif?view=net-8.0#system-objectdisposedexception-throwif(system-boolean-system-object)

**Comment by BDisp** [comment]
That will be a larger changes. I'm curious to see this :-)

**Comment by dodexahedron** [comment]
Might sound like it, but really it's one line per property accessor, and one private instance boolean field for each type. It's a safe copy/paste job, too, since there's no variation for how it is invoked anywhere, and it only really needs to be put in publicly visible members.

**Comment by dodexahedron** [comment]
Oh, and any events on an IDisposable type should immediately abort before invoking their delegates, if the object has been disposed (can optionally remove them, too, at that time).

**Comment by dodexahedron** [comment]
In any case, the addition of that Disposing event you added in your PR already is a big improvement for consumers. Thanks for that! :)

**Comment by BDisp** [comment]
How about creating a Attribute Disposable which will throw for every properties and events that was disposed?

**Comment by dodexahedron** [comment]
That'd be just as many lines for the attributes (1 per property) and then also the attribute class, so it would be a little more work.

Plus that'd be non-standard. They made that shortcut function precisely to minimize things in a clean and consistent way.

I like attributes for plenty of things and actually have an idea brewing for a proposal to help reduce boilerplate and duplicated effort between Terminal.Gui and TerminalGuiDesigner, but that's an idea that's still forming and of course off-topic here. 😅

**Comment by BDisp** [comment]
I was wrong about using `Attribute` because they are used at compile time and we cannot used an object instance to inspect the disposable value. So, the solution is throwing on each property the `ObjectDisposedException` if it was been disposed.

**Comment by dodexahedron** [comment]
Correct. Attributes are metadata only, and custom attributes do not/cannot affect program execution unless explicitly searched for and acted upon by the consuming code.

**Comment by BDisp** [comment]
Another complexity isn't only the fact that the setter have to throws the exception but also the getter when reading his value, or I'm wrong?

**Comment by dodexahedron** [comment]
Only when absolutely necessary.

Per the [property design guidelines](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/property) 👍 

>❌ AVOID throwing exceptions from property getters.
> Property getters should be simple operations and should not have any preconditions. If a getter can throw an exception, it should probably be redesigned to be a method. Notice that this rule does not apply to indexers, where we do expect exceptions as a result of validating the arguments.
